homeassistant/components/zwave_js/__init__.py

```diff
@@ -624,6 +629,46 @@ def register_node_in_dev_reg(self, node: ZwaveNode) -> dr.DeviceEntry:
         else:
             ids = {device_id}
 
+        # Check if this node was previously provisioned
+        provisioning_entry = await controller.async_get_provisioning_entry(node.node_id)
+
+        if (
+            provisioning_entry
+            and hasattr(provisioning_entry, "device_id")
+            and provisioning_entry.device_id
+        ):
+            preprovisioned_device = self.dev_reg.async_get(provisioning_entry.device_id)
+
+            if preprovisioned_device:
+                dsk = provisioning_entry.dsk
+                dsk_identifier = (DOMAIN, f"provision_{dsk}")
+
+                # If the pre-provisioned device has the DSK identifier, remove it
+                if dsk_identifier in preprovisioned_device.identifiers:
+                    new_identifiers = preprovisioned_device.identifiers.copy()
+                    new_identifiers.remove(dsk_identifier)
+                    new_identifiers.update(ids)
+                    via_device_id = None
+                    if via_identifier:
+                        # async_update_device expects a device id not a device identifier
+                        via_device = self.dev_reg.async_get_device(
+                            identifiers={via_identifier}
+                        )
+                        if via_device:
+                            via_device_id = via_device.id
+                    self.dev_reg.async_update_device(
+                        preprovisioned_device.id,
+                        new_identifiers=new_identifiers,
+                        sw_version=node.firmware_version,
+                        name=node.name
+                        or node.device_config.description
+                        or f"Node {node.node_id}",
+                        model=node.device_config.label,
+                        manufacturer=node.device_config.manufacturer,
```

Comment:
Model and manufacturer are hopefully the same, so we may not need to update them, although it may not hurt either. Not sure.